Add KEP for Cron Horizontal Pod Autoscaler(CronHPA)

hex108 opened the pull request:
We have implemented a CronHPA which enables us to scale workloads(those support `scale` subresource) periodically using crontab scheme.

Use cases
A lot of game player will play games from Friday evening to Sunday evening. It will be better to provide a better experience for players if game servers(pods) could be scaled to a lager size at Friday evening and scaled to normal size at Sunday evening. And that's what game server admins do every week.

Some other customers also do similar things because their products' usage also have peaks and valleys periodically. CronHPA is for their requirement.

CronHPA example:
```
---
apiVersion: gs.io/v1alpha1
kind: CronHPA
metadata:
  name: example-cron-hpa
spec:
  scaleTargetRef:
    apiVersion: extensions/v1beta1
    kind: Deployment
    name: demo-deployment
  crons:
    - schedule: "0 23 * * 5"  // Set replicas to 60 every Friday 23:00
      targetReplicas: 60
    - schedule: "0 23 * * 7"  // Set replicas to 30 every Sunday 23:00
      targetReplicas: 30

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: demo-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
        - containerPort: 80
```

hex108 commented:
Moved from https://github.com/kubernetes/community/pull/2976

/sig autoscaling
/assign @DirectXMan12

hex108 commented:
The code is open-sourced at https://github.com/hex108/cron-hpa-controller in case someone is interested it. :)

wojtek-t commented:
/assign @MaciekPytel 
@kubernetes/sig-autoscaling-feature-requests

mattfarina commented on keps/sig-autoscaling/0033-20181203-cron-hpa.md:
```diff
@@ -0,0 +1,183 @@
+  name: example-cron-hpa
+spec:
+  scaleTargetRef:
+    apiVersion: extensions/v1beta1
```
Deployment and the other apps are now at `apps/v1` instead of `extensions/v1beta1`. Can you update the references to the current versions?

hex108 replied:
Addressed. Thanks!

ringtail commented:
@DirectXMan12 @hex108 Will this cron hpa controller be merged to hpa controller or as a custom CRD controller? We developed a CRD controller almost the same as `hex108/cron-hpa-controller` and   being used in production for a few months. So maybe I cam merge some features to the official repo later.

hex108 commented on keps/sig-autoscaling/0033-20181203-cron-hpa.md:
```diff
@@ -0,0 +1,183 @@
+reviewers:
+  - "@DirectXMan12"
+approvers:
+  - "@DirectXMan12"
```
Addressed. Thanks!

hex108 commented:
Just addressed above comments. The design doc is sent at https://groups.google.com/forum/#!topic/kubernetes-sig-autoscaling/nxHpTF2bAyk. Thanks @mattfarina !

@ringtail Glad to see that you need this feature too! Let's wait for more input from @DirectXMan12 , @MaciekPytel, @mattfarina, @wojtek-t  and others, thanks!

mattfarina commented:
I see this landed on the SIG Autoscaling mailing list at https://groups.google.com/forum/#!topic/kubernetes-sig-autoscaling/nxHpTF2bAyk

@hex108 I would suggest 3 things...

1. Try to get this into a SIG Autoscaling meeting, as a discussion topic, when one happens again
2. Attempt to get some time with @mwielgus,  the chair of SIG Autoscaling, to discuss
3. Make this a GA quality and standalone project outside of Kubernetes

I suggest point 3 because:

* It will be in better shape if it is pulled in
* The current general guidance has been to implement controllers/objects like this in the ecosystem (leveraging CRDs) rather than as a core part of Kubernetes. Anything that lands in Kubernetes itself tends to be an exception.

This is how I would approach it if I were in your shoes

hex108 commented:
Thanks @mwielgus for the suggestions!

I prefer option C because it is more generic. In CronHPA it will only change instance number at time T, however if we implemented it in HPA, we need continuously generate  appropriate metric since time T, otherwise HPA will decrease instance number once there is no metrics, I think it might be a problem.

hex108 commented:
Thanks @mwielgus for your suggestion, close it now.

hex108 commented:
/close
